src/Cli/dotnet/Commands/Run/VirtualProjectBuildingCommand.cs

```diff
@@ -847,7 +847,28 @@ public static void RemoveDirectivesFromFile(ImmutableArray<CSharpDirective> dire
 
     public static bool IsValidEntryPointPath(string entryPointFilePath)
     {
-        return entryPointFilePath.EndsWith(".cs", StringComparison.OrdinalIgnoreCase) && File.Exists(entryPointFilePath);
+        if (!File.Exists(entryPointFilePath))
```

Comment:
I am interested to know what happens when command `dotnet build` is run, when the working directory contains a file `build` which has a `#!`, etc.

Reply:
This PR doesn't handle the `dotnet <file[.cs]>` case, you still have to do `dotnet run <file[.cs]>` right now, so I guess we'll define that behavior when we do that side of the feature. I believe it's tracked separately under: https://github.com/dotnet/sdk/issues/49202

Reply:
nose goes, whoever merges second between this and #48387 gets to deal with it.

Reply:
the built-in build command should take precedence (in fact, everything should take precedence, invoking the file is the last fallback), but I can definitely add a test in my PR (or in a follow up) once this one goes in